lib/membrane/video_compositor/video_compositor.ex

```diff
@@ -0,0 +1,431 @@
+    accepted_format: %Membrane.H264{alignment: :nalu, stream_structure: :annexb},
+    availability: :on_request,
+    options: [
+      input_id: [
+        spec: input_id(),
+        description: "Input identifier."
+      ]
+    ]
+
+  def_output_pad :output,
+    accepted_format: %Membrane.H264{alignment: :nalu, stream_structure: :annexb},
+    availability: :on_request,
+    options: [
+      resolution: [
+        spec: Resolution.t(),
+        description: "Resolution of output stream."
+      ],
+      output_id: [
+        spec: output_id(),
+        description: "Output identifier."
+      ],
+      encoder_preset: [
+        spec: encoder_preset(),
+        description:
+          "Preset for an encoder. See [FFmpeg docs](https://trac.ffmpeg.org/wiki/Encode/H.264#Preset) to learn more.",
+        default: :medium
+      ]
+    ]
+
+  @impl true
+  def handle_init(_ctx, opt) do
+    vc_port = opt.vc_server_port_number
+    :ok = start_vc_server(vc_port)
+
+    :ok = VcReq.init(opt.framerate, opt.stream_fallback_timeout, opt.init_web_renderer?, vc_port)
+
+    if opt.start_composing_strategy == :on_init do
+      :ok = VcReq.start_composing(vc_port)
+    end
+
+    {[],
+     %State{
+       inputs: [],
+       outputs: [],
+       framerate: opt.framerate,
+       vc_port: vc_port
+     }}
+  end
+
+  @impl true
+  def handle_pad_added(input_ref = Pad.ref(:input, pad_id), ctx, state = %State{inputs: inputs}) do
+    input_id = ctx.options.input_id
+    input_port = register_input_stream(input_id, state)
+
+    state = %State{
+      state
+      | inputs: [
+          %InputState{input_id: input_id, port_number: input_port, pad_ref: input_ref} | inputs
+        ]
+    }
+
+    spec =
+      bin_input(Pad.ref(:input, pad_id))
+      |> via_in(Pad.ref(:input, pad_id),
+        options: [payloader: RTP.H264.Payloader]
+      )
+      |> child({:rtp_sender, pad_id}, RTP.SessionBin)
+      |> via_out(Pad.ref(:rtp_output, pad_id), options: [encoding: :H264])
+      |> child({:upd_sink, pad_id}, %UDP.Sink{
+        destination_port_no: input_port,
+        destination_address: @local_host
+      })
+
+    {[notify_parent: {:input_registered, input_ref, input_id, State.ctx(state)}, spec: spec],
+     state}
+  end
+
+  @impl true
+  def handle_pad_added(
+        output_ref = Pad.ref(:output, pad_id),
+        ctx,
+        state = %State{outputs: outputs}
+      ) do
+    port = register_output_stream(ctx.options, state)
+    output_id = ctx.options.output_id
+
+    state = %State{
+      state
+      | outputs: [
+          %OutputState{
+            output_id: output_id,
+            pad_ref: output_ref,
+            port_number: port,
+            resolution: ctx.options.resolution
+          }
+          | outputs
+        ]
+    }
+
+    spec =
+      child(Pad.ref(:upd_source, pad_id), %UDP.Source{
+        local_port_no: port,
+        local_address: @local_host,
+        recv_buffer_size: @udp_buffer_size
+      })
```

Comment:
Using `Pad.ref` to name a child is weird, please change it 🙈 🙈 🙈 You can name this child just `{:udp_source, pad_id}`